clang/lib/CodeGen/CGExprComplex.cpp

```diff
@@ -794,8 +834,10 @@ ComplexPairTy ComplexExprEmitter::EmitBinMul(const BinOpInfo &Op) {
       ResR = Builder.CreateFSub(AC, BD, "mul_r");
       ResI = Builder.CreateFAdd(AD, BC, "mul_i");
 
-      if (Op.FPFeatures.getComplexRange() == LangOptions::CX_Limited ||
-          Op.FPFeatures.getComplexRange() == LangOptions::CX_Fortran)
+      if (Op.FPFeatures.getComplexRange() == LangOptions::CX_Basic ||
+          Op.FPFeatures.getComplexRange() == LangOptions::CX_Improved ||
+          Op.FPFeatures.getComplexRange() == LangOptions::CX_Promoted ||
+          CGF.getLangOpts().NoHonorInfs || CGF.getLangOpts().NoHonorNaNs)
```

Comment:
I'm not sure NoHonorInfs or NoHonorNaNs should be checked here. Given that we have explicit control for complex arithmetic behavior, maybe that should take precedence. That seems to be the way gcc handles it: https://godbolt.org/z/1oGo7jznz

Reply:
This code is in `EmitBinMul` and your `godbolt` link has a `div` operation.
At any case the `complex-range` is **not** set when `-ffinite-math-only` is used on the command line. The function `applyFastMath` sets the range only when `ffast-math` or `fp-model` are used. With `CGF.getLangOpts().NoHonorInfs || CGF.getLangOpts().NoHonorNaNs` we would get the same behavior than `gcc` for a` mul` operation: algebraic  implementation with no `nan` testing.
If you don't want this condition  here. then `applyFastMath` needs to be called when `ffinite-math-only` is on the command line.
WDYT?

Reply:
Oh, sorry, for some reason I thought this was invoking the special handling for division. I think the for honor NaNs and infinities still isn't necessary because we'll emit the comparison with the fast-math flags set and the backend will optimize it away, which is what happens today: https://godbolt.org/z/e8EnKodj6